### ISSUE TITLE
Design issues how to convert ASR->Fortran

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#2746]
There are several issues that we need to think about.

# Underscore in variables

Automatically generated variables in ASR often look like `__1_t` and on purpose contain characters that are not valid Fortran or Python variable names (in this case they are invalid in Fortran, but valid in Python; however sometimes we also use `@`, then it's invalid in both). 

At the same time, as a competing requirement, we want the generated Fortran code to be executable with gfortran and other Fortran compilers.

How to resolve this? 

* One way is to make this configurable, and add a compiler option to rewrite names. However, any ASR rewrite is often tricky. 
* We could add a compiler pass that renames variables to be valid Fortran variables

I like the second approach the most so far. That way we print exactly what is in ASR and maybe it is not 100% compilable with Fortran, but it's still good for visual inspection and debugging. And we can provide some passes that one can optionally call to post process things to be compilable. Then `--show-fortran` can automatically call them, but `--dump-all-passes-fortran` will not call them.

**Comment by Thirumalai-Shaktivel** [comment]
I thought of prepending "v" for variables and "f" for functions in the backend itself. Would this cause any problems?
`__i` -> `v__i` and `_compilers_fn` -> `f_compilers_fn`

**Comment by Shaikh-Ubaid** [comment]
Right, I have faced these issues. Is it possible to simply use names in the "Automatically generated variables in ASR" that are compatible to most compilers like gfortran, python, c++ and more. 

I think most compilers would support `_`, alphabets and numbers and the starting character should be a character.

>We could add a compiler pass that renames variables to be valid Fortran variables

This is possible, but a pass in my opinion iterates over the whole code thus affecting the compilation time.

**Comment by Thirumalai-Shaktivel** [comment]
Something like this: https://github.com/lfortran/lfortran/commit/392ca583e2d36f7db3ac463e5da6ad4fba3641be

**Comment by Shaikh-Ubaid** [comment]
PS: It seems we also have a `unique_symbols` pass which helps to mangle names. Maybe we could simply use this pass and make it mangle names such that they become valid fortran variables.

**Comment by certik** [comment]
Yes, I recommend to use passes for "Fortran backend" to cleanup the ASR so that GFortran can compile it. For "Fortran visualization" I would generate ASR directly, and we can teach LFortran to compile it (perhaps with some extra compiler options).

We'll have to eventually resolve what to do about variables like `@internal@i`. We have to "escape" them somehow, or not allow them in ASR in the first place.

**Comment by Shaikh-Ubaid** [comment]
>or not allow them in ASR in the first place.

I am on the side of not allowing them in the first place. I think that would be the most complete/final fix for the longer term. Also, it seems unusual/odd to see such names (`@internal@i`) for variables/symbols when reading ASR while debugging issues. I would simply use compiler friendly names using a combination of capital and small alphabets, numbers and (if most compilers support) underscore (except for the start character).

The above is just a suggestion. Any approach is fine with me.

**Comment by certik** [comment]
The reason to use `@` was to disambiguate it from user symbols, to ensure they do not clash. However I like the conceptual "academic" design that ASR is just Fortran and Python code, so it should use such variable names that work in both languages and we already have functions to give us a new unique symbol name.

However, there is another problem: Fortran is case insensitive, so `a` and `A` are the same variable, while in Python they are two different variables. In ASR we use case sensitive approach (currently we convert all Fortran variables to lowercase). However, when translating Python code that will have both `a` and `A`, how should that be handled in Fortran?
